[a11y] : set isFocused will update isFocusable to true

hannah-hyj opened the pull request:
the isFocused and isFocusable flags should be more consistent,

this is also to prepare for https://github.com/flutter/flutter/pull/170696 ,


So all EditableText and TextField widgets will have isFocusable flag set to true now. (because they set isFocused).



## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -5453,9 +5453,11 @@ class SemanticsConfiguration {
-  bool get isFocused => _flags.isFocused;
-  set isFocused(bool value) {
-    _flags = _flags.copyWith(isFocused: value);
+  bool? get isFocused => _flags.isFocusable ? _flags.isFocused : null;
```
consider marking isFocusable deprecated

hannah-hyj commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -5446,7 +5446,16 @@ class SemanticsConfiguration {
 
   /// Whether the owning [RenderObject] can hold the input focus.
+  @Deprecated(
+    'Use isFocused instead. '
```
they can see if isFocused is null?

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -5446,16 +5446,27 @@ class SemanticsConfiguration {
 
   /// Whether the owning [RenderObject] can hold the input focus.
+  @Deprecated(
+    'Check if isFocused is null instead. '
```
oh i didn't notice isFocused is tristate lgtm

chunhtai commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -5446,16 +5446,27 @@ class SemanticsConfiguration {
-    _flags = _flags.copyWith(isFocused: value);
+  bool? get isFocused => _flags.isFocusable ? _flags.isFocused : null;
+  set isFocused(bool? value) {
+    if (value != null) {
```
if this is null, should this reset isFocusable? I looked at other property looks like they just pass the null to the flags, not sure if those make sense neither

chunhtai commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -955,7 +955,7 @@ class RenderCustomPaint extends RenderProxyBox {
       config.isFocusable = properties.focusable!;
```
we should attempt to migrate this, the focusable in SemanticsProperty should be deprecated as well

chunhtai reviewed (commented):
we also need to add a migration guide

chunhtai commented on packages/flutter/lib/src/widgets/focus_scope.dart:
```diff
@@ -726,7 +726,7 @@ class _FocusState extends State<Focus> {
                 ? focusNode.requestFocus
                 : null,
         focusable: _couldRequestFocus,
```
looks like Semantics(focusable) is not deprecated. is there a reason we want to keep it around?

hannah-hyj replied:
Yeah i should remove this, my mistake

hannah-hyj replied:
Actually removing this line caused a lot of test failures in g3, 
When Semantics(focusable: false, focused: false) becomes Semantics(focused: null), they are the same in flag values. However, the first one is **annotated** in the semantics tree, while the second one is not. This difference can sometimes cause the tree structure to change.

Because Focus is so widely used, it caused many tests to fail. For example, an InkWell(onTap: null) might have previously formed a semantics node because it was annotated, but with focusable removed here, it will no longer form a semantics node.

I agree I should remove the invocation of `focusable` here but i will split this 1 line change into another PR like i split the textfield focusable into another PR https://github.com/flutter/flutter/pull/173235 ,  just so it's easier to resolve semantics tree structure changes and test failures.

hannah-hyj commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -951,11 +951,8 @@ class RenderCustomPaint extends RenderProxyBox {
-    }
-    if (properties.focused != null) {
-      config.isFocused = properties.focused!;
+    if (properties.focusable != null ||properties.focused != null) {
```
discussed offline, the tree structure changes is intended

chunhtai reviewed (approved):
LGTM

hannah-hyj commented:
I'm splitting this PR into https://github.com/flutter/flutter/pull/173234 and this so it will be easier to debug why google3 tests are failing.

hannah-hyj committed:
Update dropdown_menu_test.dart

1

isfocusable changes

revert inkwell changes

isfocusable changes

Update ink_well.dart

Update semantics.dart

lint

Update shortcuts_test.dart

fix tests

test

resolve comments

fix tests

Update semantics.dart

fix tests

resolve comments

Update semantics.dart

Update packages/flutter/lib/src/semantics/semantics.dart

mark Deprecated

Update semantics_test.dart

update more tests

fix tests

updates

lint

Update semantics.dart

Update semantics.dart

Update semantics.dart

Co-Authored-By: chunhtai <47866232+chunhtai@users.noreply.github.com>

hannah-hyj committed:
revert changes in focus_scope revert tests

Update menu_anchor_test.dart

revert changes in focus_scope

Co-Authored-By: chunhtai <47866232+chunhtai@users.noreply.github.com>

hannah-hyj committed:
version number

Co-Authored-By: chunhtai <47866232+chunhtai@users.noreply.github.com>
